sandbox/lua/decoders/syslog.lua

```diff
@@ -0,0 +1,373 @@
+                 * capture_followed_by('pam_type', '): ')
+wildcard_grammar['pam'] = l.Ct(
+                        (   pam_header
+                          * l.P'session opened for user '
+                          * capture_followed_by('user_name', ' by ')
+                          * capture_followed_by('login_name', '(uid=')
+                          * l.Cg(l.digit^1 / tonumber, 'uid')
+                          * l.P')'
+                        ) + (
+                            pam_header
+                          * l.P'session closed for user '
+                          * l.Cg(l.P(1)^1, 'user_name')
+                        ) + (
+                            pam_header
+                          * l.P'authentication failure; logname='
+                          * capture_followed_by('logname', ' uid=')
+                          * l.Cg(l.digit^1 / tonumber, 'uid')
+                          * l.P' euid='
+                          * l.Cg(l.digit^1 / tonumber, 'euid')
+                          * l.P' tty='
+                          * capture_followed_by('tty', ' ruser=')
+                          * capture_followed_by('ruser', ' rhost=')
+                          * l.Cg((l.P(1) - l.P(' user='))^0, 'rhost')
+                          * (l.P' user=' * l.Cg(l.P(1)^1, 'user'))^-1
+                        ) + (
+                            pam_header
+                          * l.P'check pass; user '
+                          * (l.P'(' * capture_followed_by('user_name', ') '))^-1
+                          * l.P'unknown'
+                        ))
+
+-- end of FIXME
+
+function process_message ()
+    local log = read_message("Payload")
+    local fields = nil
+    if rsyslog_grammar then
+        fields = rsyslog_grammar:match(log)
+        if not fields then return -1 end
+        if fields.timestamp then
+            msg.Timestamp = fields.timestamp
+            fields.timestamp = nil
+        end
+
+        if fields.pri then
+            msg.Severity = fields.pri.severity
+            fields.syslogfacility = fields.pri.facility
+            fields.pri = nil
+        else
+            msg.Severity = fields.syslogseverity or fields["syslogseverity-text"]
+            or fields.syslogpriority or fields["syslogpriority-text"]
+
+            fields.syslogseverity = nil
+            fields["syslogseverity-text"] = nil
+            fields.syslogpriority = nil
+            fields["syslogpriority-text"] = nil
+        end
+
+        if fields.syslogtag then
+            fields.programname = fields.syslogtag.programname
+            msg.Pid = fields.syslogtag.pid
+            fields.syslogtag = nil
+        end
+
+        if not hostname_keep then
+            msg.Hostname = fields.hostname or fields.source
+            fields.hostname = nil
+            fields.source = nil
+        end
+
+        msg.Payload = fields.msg
+        fields.msg = nil
+
+    else
+        msg = decode_message(read_message('raw'))
```

Comment:
Hrm, I'm a bit confused about this. `decode_message(read_message('raw'))` is for cases when we want to deserialize the protobuf encoded message bytes that are stored on the pack. Not sure why we'd want that to happen if no rsyslog template is provided... wouldn't we still be expecting to use the text data from the message payload?